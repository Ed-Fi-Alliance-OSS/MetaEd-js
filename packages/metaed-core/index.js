// @flow

// Test helper -- *** where should this go? ***
export { MetaEdTextBuilder } from './test/MetaEdTextBuilder';


// Functions

export { deepFreeze, deepFreezeAssign, prependIndefiniteArticle } from './src/Utility';
export { startingFromFileLoadP } from './src/task/Pipeline';
export { createMetaEdFile } from './src/task/MetaEdFile';
export { loadCoreBufferedFiles, loadExtensionBufferedFiles } from './src/task/BufferFileLoader';
export { loadFiles } from './src/task/FileSystemFilenameLoader';
export { newState } from './src/State';
export { addEntity, getAllEntities, getAllTopLevelEntities, getEntity, getTopLevelCoreEntity, getEntitiesOfType, getAllEntitiesNoSimpleTypes } from './src/model/EntityRepository';
export { addProperty, getAllProperties, getPropertiesOfType, newPropertyIndex } from './src/model/property/PropertyRepository';
export { isReferenceProperty, isSharedProperty, allPropertyTypes } from './src/model/property/PropertyType';
export { allEntityModelTypes, allTopLevelEntityModelTypes, topLevelCoreEntityModelTypes } from './src/model/ModelType';

// Builders

export { AssociationBuilder } from './src/builder/AssociationBuilder';
export { AssociationExtensionBuilder } from './src/builder/AssociationExtensionBuilder';
export { AssociationSubclassBuilder } from './src/builder/AssociationSubclassBuilder';
export { ChoiceBuilder } from './src/builder/ChoiceBuilder';
export { CommonBuilder } from './src/builder/CommonBuilder';
export { CommonExtensionBuilder } from './src/builder/CommonExtensionBuilder';
export { DecimalTypeBuilder } from './src/builder/DecimalTypeBuilder';
export { DescriptorBuilder } from './src/builder/DescriptorBuilder';
export { DomainBuilder } from './src/builder/DomainBuilder';
export { DomainEntityBuilder } from './src/builder/DomainEntityBuilder';
export { DomainEntityExtensionBuilder } from './src/builder/DomainEntityExtensionBuilder';
export { DomainEntitySubclassBuilder } from './src/builder/DomainEntitySubclassBuilder';
export { EnumerationBuilder } from './src/builder/EnumerationBuilder';
export { IntegerTypeBuilder } from './src/builder/IntegerTypeBuilder';
export { InterchangeBuilder } from './src/builder/InterchangeBuilder';
export { NamespaceInfoBuilder } from './src/builder/NamespaceInfoBuilder';
export { SharedDecimalBuilder } from './src/builder/SharedDecimalBuilder';
export { SharedIntegerBuilder } from './src/builder/SharedIntegerBuilder';
export { SharedStringBuilder } from './src/builder/SharedStringBuilder';
export { StringTypeBuilder } from './src/builder/StringTypeBuilder';

// Flow types

export type { State } from './src/State';
export type { InputDirectory } from './src/task/FileSystemFilenameLoader';
export type { FileSet, MetaEdFile } from './src/task/MetaEdFile';
export type { FileIndex } from './src/task/FileIndex';
export type { Validator } from './src/validator/Validator';
export type { ValidationFailure } from './src/validator/ValidationFailure';
export type { Enhancer } from './src/enhancer/Enhancer';
export type { EnhancerResult } from './src/enhancer/EnhancerResult';
export type { Generator } from './src/generator/Generator';
export type { GeneratorResult } from './src/generator/GeneratorResult';
export type { GeneratedOutput } from './src/generator/GeneratedOutput';
export type { MetaEdEnvironment, PluginEnvironment, SemVer } from './src/MetaEdEnvironment';
export type { MetaEdPlugin, PluginManifest } from './src/plugin/PluginTypes';

// Properties
export type { AssociationProperty } from './src/model/property/AssociationProperty';
export type { BooleanProperty } from './src/model/property/BooleanProperty';
export type { ChoiceProperty } from './src/model/property/ChoiceProperty';
export type { CommonProperty } from './src/model/property/CommonProperty';
export type { CurrencyProperty } from './src/model/property/CurrencyProperty';
export type { DateProperty } from './src/model/property/DateProperty';
export type { DecimalProperty } from './src/model/property/DecimalProperty';
export type { DescriptorProperty } from './src/model/property/DescriptorProperty';
export type { DomainEntityProperty } from './src/model/property/DomainEntityProperty';
export type { DurationProperty } from './src/model/property/DurationProperty';
export type { EntityProperty, HasReferencedEntity } from './src/model/property/EntityProperty';
export type { EnumerationProperty } from './src/model/property/EnumerationProperty';
export type { InlineCommonProperty } from './src/model/property/InlineCommonProperty';
export type { IntegerProperty } from './src/model/property/IntegerProperty';
export type { MergedProperty } from './src/model/property/MergedProperty';
export type { PercentProperty } from './src/model/property/PercentProperty';
export type { PropertyIndex } from './src/model/property/PropertyRepository';
export type { PropertyType } from './src/model/property/PropertyType';
export type { ReferentialProperty } from './src/model/property/ReferentialProperty';
export type { SchoolYearEnumerationProperty } from './src/model/property/SchoolYearEnumerationProperty';
export type { SharedDecimalProperty } from './src/model/property/SharedDecimalProperty';
export type { SharedIntegerProperty } from './src/model/property/SharedIntegerProperty';
export type { SharedShortProperty } from './src/model/property/SharedShortProperty';
export type { SharedStringProperty } from './src/model/property/SharedStringProperty';
export type { ShortProperty } from './src/model/property/ShortProperty';
export type { SimpleProperty } from './src/model/property/SimpleProperty';
export type { SharedSimple } from './src/model/SharedSimple';
export type { StringProperty } from './src/model/property/StringProperty';
export type { TimeProperty } from './src/model/property/TimeProperty';
export type { YearProperty } from './src/model/property/YearProperty';

// Entities
export type { AbstractEntity } from './src/model/AbstractEntity';
export type { Association } from './src/model/Association';
export type { AssociationExtension } from './src/model/AssociationExtension';
export type { AssociationSubclass } from './src/model/AssociationSubclass';
export type { Choice } from './src/model/Choice';
export type { Common } from './src/model/Common';
export type { CommonExtension } from './src/model/CommonExtension';
export type { DecimalType } from './src/model/DecimalType';
export type { Descriptor } from './src/model/Descriptor';
export type { Domain } from './src/model/Domain';
export type { DomainEntity } from './src/model/DomainEntity';
export type { DomainEntityExtension } from './src/model/DomainEntityExtension';
export type { DomainEntitySubclass } from './src/model/DomainEntitySubclass';
export type { DomainItem } from './src/model/DomainItem';
export type { EntityRepository } from './src/model/EntityRepository';
export type { Enumeration } from './src/model/Enumeration';
export type { EnumerationItem } from './src/model/EnumerationItem';
export type { IntegerType } from './src/model/IntegerType';
export type { Interchange } from './src/model/Interchange';
export type { InterchangeExtension } from './src/model/InterchangeExtension';
export type { InterchangeItem } from './src/model/InterchangeItem';
export type { MapTypeEnumeration } from './src/model/MapTypeEnumeration';
export type { ModelBase } from './src/model/ModelBase';
export type { ModelType } from './src/model/ModelType';
export type { NamespaceInfo } from './src/model/NamespaceInfo';
export type { SchoolYearEnumeration } from './src/model/SchoolYearEnumeration';
export type { SharedDecimal } from './src/model/SharedDecimal';
export type { SharedInteger } from './src/model/SharedInteger';
export type { SharedString } from './src/model/SharedString';
export type { SourceMap } from './src/model/SourceMap';
export type { StringType } from './src/model/StringType';
export type { Subdomain } from './src/model/Subdomain';
export type { TopLevelEntity } from './src/model/TopLevelEntity';

// Null Objects
export { NoMapTypeEnumeration } from './src/model/MapTypeEnumeration';
export { NoTopLevelEntity } from './src/model/TopLevelEntity';

// Factories
export { newMetaEdEnvironment, newPluginEnvironment } from './src/MetaEdEnvironment';
export { newMetaEdPlugin } from './src/plugin/PluginTypes';

export { newAssociationProperty } from './src/model/property/AssociationProperty';
export { newBooleanProperty } from './src/model/property/BooleanProperty';
export { newChoiceProperty } from './src/model/property/ChoiceProperty';
export { newCommonProperty } from './src/model/property/CommonProperty';
export { newCurrencyProperty } from './src/model/property/CurrencyProperty';
export { newDateProperty } from './src/model/property/DateProperty';
export { newDecimalProperty } from './src/model/property/DecimalProperty';
export { newDescriptorProperty } from './src/model/property/DescriptorProperty';
export { newDomainEntityProperty } from './src/model/property/DomainEntityProperty';
export { newDurationProperty } from './src/model/property/DurationProperty';
export { newEnumerationProperty } from './src/model/property/EnumerationProperty';
export { newInlineCommonProperty } from './src/model/property/InlineCommonProperty';
export { newIntegerProperty } from './src/model/property/IntegerProperty';
export { newMergedProperty } from './src/model/property/MergedProperty';
export { newPercentProperty } from './src/model/property/PercentProperty';
export { newSchoolYearEnumerationProperty } from './src/model/property/SchoolYearEnumerationProperty';
export { newSharedDecimalProperty } from './src/model/property/SharedDecimalProperty';
export { newSharedIntegerProperty } from './src/model/property/SharedIntegerProperty';
export { newSharedShortProperty } from './src/model/property/SharedShortProperty';
export { newSharedStringProperty } from './src/model/property/SharedStringProperty';
export { newShortProperty } from './src/model/property/ShortProperty';
export { newSimpleProperty } from './src/model/property/SimpleProperty';
export { newStringProperty } from './src/model/property/StringProperty';
export { newTimeProperty } from './src/model/property/TimeProperty';
export { newYearProperty } from './src/model/property/YearProperty';

export { newAbstractEntity } from './src/model/AbstractEntity';
export { newAssociation } from './src/model/Association';
export { newAssociationExtension } from './src/model/AssociationExtension';
export { newAssociationSubclass } from './src/model/AssociationSubclass';
export { newChoice } from './src/model/Choice';
export { newCommon } from './src/model/Common';
export { newCommonExtension } from './src/model/CommonExtension';
export { newDecimalType } from './src/model/DecimalType';
export { newDescriptor } from './src/model/Descriptor';
export { newDomain } from './src/model/Domain';
export { newDomainEntity } from './src/model/DomainEntity';
export { newDomainEntityExtension } from './src/model/DomainEntityExtension';
export { newDomainEntitySubclass } from './src/model/DomainEntitySubclass';
export { newDomainItem } from './src/model/DomainItem';
export { newEnumeration } from './src/model/Enumeration';
export { newEnumerationItem } from './src/model/EnumerationItem';
export { newInlineCommon } from './src/model/Common';
export { newIntegerType } from './src/model/IntegerType';
export { newInterchange } from './src/model/Interchange';
export { newInterchangeExtension } from './src/model/InterchangeExtension';
export { newInterchangeItem } from './src/model/InterchangeItem';
export { newMapTypeEnumeration } from './src/model/MapTypeEnumeration';
export { newNamespaceInfo } from './src/model/NamespaceInfo';
export { newSchoolYearEnumeration } from './src/model/SchoolYearEnumeration';
export { newSharedDecimal } from './src/model/SharedDecimal';
export { newSharedInteger } from './src/model/SharedInteger';
export { newSharedString } from './src/model/SharedString';
export { newSourceMap } from './src/model/SourceMap';
export { newStringType } from './src/model/StringType';
export { newSubdomain } from './src/model/Subdomain';
export { asTopLevelEntity, newTopLevelEntity } from './src/model/TopLevelEntity';
