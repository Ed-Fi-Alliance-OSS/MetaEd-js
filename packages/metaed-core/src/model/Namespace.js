// @flow
import deepFreeze from 'deep-freeze';
import type { SourceMap } from './SourceMap';
import { NoSourceMap } from './SourceMap';
import type { ModelType } from './ModelType';

export type NamespaceSourceMap = {
  type: SourceMap,
  namespaceName: SourceMap,
  isExtension: SourceMap,
  projectExtension: SourceMap,
  projectName: SourceMap,
  extensionEntitySuffix: SourceMap,
};

export function newNamespaceSourceMap(): NamespaceSourceMap {
  return {
    type: NoSourceMap,
    namespaceName: NoSourceMap,
    isExtension: NoSourceMap,
    projectExtension: NoSourceMap,
    projectName: NoSourceMap,
    extensionEntitySuffix: NoSourceMap,
  };
}

export type Namespace = {
  type: ModelType,
  namespaceName: string,
  isExtension: boolean,
  projectExtension: string,
  projectName: string,
  extensionEntitySuffix: string,
  sourceMap: NamespaceSourceMap,
  data: any,
  config: any,
};

export const DefaultExtensionEntitySuffix: string = 'Extension';

export function newNamespace(): Namespace {
  return {
    type: 'namespace',
    namespaceName: '',
    isExtension: false,
    projectExtension: '',
    projectName: '',
    extensionEntitySuffix: DefaultExtensionEntitySuffix,
    sourceMap: newNamespaceSourceMap(),
    data: {},
    config: {},
  };
}

export const NoNamespace: Namespace = deepFreeze({
  ...newNamespace(),
  namespaceName: 'nonamespace',
  projectExtension: 'NoNamespace',
});
