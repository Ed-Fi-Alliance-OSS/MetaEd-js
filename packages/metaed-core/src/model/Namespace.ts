import deepFreeze from 'deep-freeze';
import { EntityRepository } from './EntityRepository';
import { newEntityRepository } from './EntityRepository';

/**
 * A Namespace is the highest level of organization of entities, and maps to
 * a single MetaEd project. There is a namespace for the Data Standard
 * project and one for each extension project.
 */
export interface Namespace {
  entity: EntityRepository;
  namespaceName: string;
  isExtension: boolean;
  // Namespace dependencies, in closest dependency first order
  dependencies: Namespace[];
  projectExtension: string;
  projectName: string;
  projectVersion: string;
  projectDescription: string;
  extensionEntitySuffix: string;
  data: any;
}

/**
 *
 */
export const DefaultExtensionEntitySuffix = 'Extension';

/**
 *
 */
export function newNamespace(): Namespace {
  return {
    entity: newEntityRepository(),
    namespaceName: '',
    isExtension: false,
    dependencies: [],
    projectExtension: '',
    projectName: '',
    projectVersion: '',
    projectDescription: '',
    extensionEntitySuffix: DefaultExtensionEntitySuffix,
    data: {},
  };
}

/**
 *
 */
export const NoNamespace: Namespace = deepFreeze({
  ...newNamespace(),
  namespaceName: 'nonamespace',
  projectExtension: 'NoNamespace',
});
