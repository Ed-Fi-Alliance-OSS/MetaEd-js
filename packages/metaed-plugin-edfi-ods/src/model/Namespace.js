// @flow
import type { EnhancerResult, MetaEdEnvironment, Namespace } from 'metaed-core';
import { newSchemaContainer } from './database/SchemaContainer';
import type { SchemaContainer } from './database/SchemaContainer';

export type NamespaceEdfiOds = {
  ods_Schema: SchemaContainer,
};

const enhancerName: string = 'NamespaceSetupEnhancer';

export function addNamespaceEdfiOdsTo(namespace: Namespace) {
  if (namespace.data.edfiOds == null) namespace.data.edfiOds = {};

  Object.assign(namespace.data.edfiOds, {
    ods_Schema: newSchemaContainer(),
  });
}

export function enhance(metaEd: MetaEdEnvironment): EnhancerResult {
  metaEd.namespace.forEach((namespace: Namespace) => {
    addNamespaceEdfiOdsTo(namespace);
  });

  return {
    enhancerName,
    success: true,
  };
}
