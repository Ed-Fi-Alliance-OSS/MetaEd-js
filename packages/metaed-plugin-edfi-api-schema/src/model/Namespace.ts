import { MetaEdEnvironment, EnhancerResult, Namespace } from '@edfi/metaed-core';
import { ApiSchema, newApiSchema } from './api-schema/ApiSchema';
import { noDocument, type Document } from './OpenApiTypes';

export interface NamespaceEdfiApiSchema {
  apiSchema: ApiSchema;
  coreOpenApiSpecification: Document;
  openApiExtensionFragments: any;
}

const enhancerName = 'NamespaceSetupEnhancer';

export function addNamespaceEdfiApiSchema(namespace: Namespace) {
  if (namespace.data.edfiApiSchema == null) namespace.data.edfiApiSchema = {};

  Object.assign(namespace.data.edfiApiSchema, {
    apiSchema: newApiSchema(),
    coreOpenApiSpecification: noDocument,
    openApiExtensionFragments: {},
  });
}

export function enhance(metaEd: MetaEdEnvironment): EnhancerResult {
  metaEd.namespace.forEach((namespace: Namespace) => {
    addNamespaceEdfiApiSchema(namespace);
  });

  return {
    enhancerName,
    success: true,
  };
}
