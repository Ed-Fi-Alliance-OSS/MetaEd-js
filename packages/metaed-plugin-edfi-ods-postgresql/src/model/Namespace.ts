import { EnhancerResult, MetaEdEnvironment, Namespace, newPluginEnvironment } from '@edfi/metaed-core';
import { newSchemaContainer } from './SchemaContainer';
import { SchemaContainer } from './SchemaContainer';

export interface NamespaceEdfiOdsPostgresql {
  odsSchema: SchemaContainer;
}

const enhancerName = 'NamespaceSetupEnhancer';

export function addNamespaceEdfiOdsPostgresqlTo(namespace: Namespace) {
  if (namespace.data.edfiOdsPostgresql == null) namespace.data.edfiOdsPostgresql = {};

  Object.assign(namespace.data.edfiOdsPostgresql as NamespaceEdfiOdsPostgresql, {
    odsSchema: newSchemaContainer(),
  });
}

export function enhance(metaEd: MetaEdEnvironment): EnhancerResult {
  const edfiOdsPostgresqlPlugin = metaEd.plugin.get('edfiOdsPostgresql');
  if (edfiOdsPostgresqlPlugin == null) {
    metaEd.plugin.set('edfiOdsPostgresql', {
      ...newPluginEnvironment(),
      shortName: 'edfiOdsPostgresql',
    });
  }

  metaEd.namespace.forEach((namespace: Namespace) => {
    addNamespaceEdfiOdsPostgresqlTo(namespace);
  });

  return {
    enhancerName,
    success: true,
  };
}
