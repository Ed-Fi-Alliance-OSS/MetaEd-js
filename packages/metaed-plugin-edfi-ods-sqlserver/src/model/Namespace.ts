import { EnhancerResult, MetaEdEnvironment, Namespace, newPluginEnvironment } from 'metaed-core';
import { newSchemaContainer } from './SchemaContainer';
import { SchemaContainer } from './SchemaContainer';

export interface NamespaceEdfiOdsSqlServer {
  odsSchema: SchemaContainer;
}

const enhancerName = 'NamespaceSetupEnhancer';

export function addNamespaceEdfiOdsSqlServerTo(namespace: Namespace) {
  if (namespace.data.edfiOdsSqlServer == null) namespace.data.edfiOdsSqlServer = {};

  Object.assign(namespace.data.edfiOdsSqlServer, {
    odsSchema: newSchemaContainer(),
  });
}

export function enhance(metaEd: MetaEdEnvironment): EnhancerResult {
  const edfiOdsSqlServerPlugin = metaEd.plugin.get('edfiOdsSqlServer');
  if (edfiOdsSqlServerPlugin == null) {
    metaEd.plugin.set('edfiOdsSqlServer', {
      ...newPluginEnvironment(),
      shortName: 'edfiOdsSqlServer',
    });
  }

  metaEd.namespace.forEach((namespace: Namespace) => {
    addNamespaceEdfiOdsSqlServerTo(namespace);
  });

  return {
    enhancerName,
    success: true,
  };
}
