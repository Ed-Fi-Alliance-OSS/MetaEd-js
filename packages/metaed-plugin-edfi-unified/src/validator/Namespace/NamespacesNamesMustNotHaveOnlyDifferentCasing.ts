import { MetaEdEnvironment, ValidationFailure } from 'metaed-core';

const differOnlyInCaseFunction = (name: string, index: number, array: Array<string>) => {
  const found: string | undefined = array.find(
    (nameInFind: string, indexInFind: number) =>
      indexInFind !== index && nameInFind !== name && nameInFind.toLowerCase() === name.toLowerCase(),
  );
  return found != null;
};

export function validate(metaEd: MetaEdEnvironment): Array<ValidationFailure> {
  const failures: Array<ValidationFailure> = [];
  const namespaceNames: Array<string> = Array.from(metaEd.namespace.keys());
  const duplicateNames: Array<string> = Array.from(namespaceNames.filter(differOnlyInCaseFunction));

  duplicateNames.forEach(duplicateName => {
    failures.push({
      validatorName: 'NamespacesNamesMustNotHaveOnlyDifferentCasing',
      category: 'error',
      message: `${duplicateName} is a project name that differs only in case from another name.  Make the casing the same to indicate the same namespace or choose a different name.`,
      sourceMap: null,
      fileMap: null,
    });
  });

  return failures;
}
