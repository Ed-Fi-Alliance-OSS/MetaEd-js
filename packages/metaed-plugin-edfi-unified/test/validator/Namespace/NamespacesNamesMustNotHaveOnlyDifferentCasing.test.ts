import { newMetaEdEnvironment, MetaEdTextBuilder, NamespaceBuilder, DomainEntityBuilder } from 'metaed-core';
import { MetaEdEnvironment, ValidationFailure } from 'metaed-core';
import { validate } from '../../../src/validator/Namespace/NamespacesNamesMustNotHaveOnlyDifferentCasing';

describe('when validating namespaces are different', () => {
  const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
  let failures: Array<ValidationFailure>;

  beforeAll(() => {
    MetaEdTextBuilder.build()
      .withBeginNamespace('EdFi')
      .withStartDomainEntity('AnEntity')
      .withDocumentation('doc')
      .withBooleanProperty('Property1', 'doc', true, false)
      .withEndDomainEntity()
      .withEndNamespace()

      .withBeginNamespace('Extension', 'ProjectExtension')
      .withStartDomainEntity('AnotherEntity')
      .withDocumentation('doc')
      .withBooleanProperty('Property2', 'doc2', true, false)
      .withEndDomainEntity()
      .withEndNamespace()

      .sendToListener(new NamespaceBuilder(metaEd, []))
      .sendToListener(new DomainEntityBuilder(metaEd, []));

    failures = validate(metaEd);
  });

  it('should have no validation failures', () => {
    expect(failures).toHaveLength(0);
  });
});

describe('when validating namespaces are exactly the same', () => {
  const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
  let failures: Array<ValidationFailure>;

  beforeAll(() => {
    MetaEdTextBuilder.build()
      .withBeginNamespace('EdFi')
      .withStartDomainEntity('AnEntity')
      .withDocumentation('doc')
      .withBooleanProperty('Property1', 'doc', true, false)
      .withEndDomainEntity()
      .withEndNamespace()

      .withBeginNamespace('EdFi')
      .withStartDomainEntity('AnotherEntity')
      .withDocumentation('doc')
      .withBooleanProperty('Property2', 'doc2', true, false)
      .withEndDomainEntity()
      .withEndNamespace()

      .sendToListener(new NamespaceBuilder(metaEd, []))
      .sendToListener(new DomainEntityBuilder(metaEd, []));

    failures = validate(metaEd);
  });

  it('should have no validation failures', () => {
    expect(failures).toHaveLength(0);
  });
});

describe('when validating namespaces are different only in casing', () => {
  const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
  let failures: Array<ValidationFailure>;

  beforeAll(() => {
    MetaEdTextBuilder.build()
      .withBeginNamespace('EdFi')
      .withStartDomainEntity('AnEntity')
      .withDocumentation('doc')
      .withBooleanProperty('Property1', 'doc', true, false)
      .withEndDomainEntity()
      .withEndNamespace()

      .withBeginNamespace('EDFI')
      .withStartDomainEntity('AnotherEntity')
      .withDocumentation('doc')
      .withBooleanProperty('Property2', 'doc2', true, false)
      .withEndDomainEntity()
      .withEndNamespace()

      .sendToListener(new NamespaceBuilder(metaEd, []))
      .sendToListener(new DomainEntityBuilder(metaEd, []));

    failures = validate(metaEd);
  });

  it('should have validation failure', () => {
    expect(failures).toHaveLength(2);
    expect(failures[0].validatorName).toBe('NamespacesNamesMustNotHaveOnlyDifferentCasing');
    expect(failures[0].category).toBe('error');
    expect(failures[0].message).toMatchSnapshot();
    expect(failures[0].sourceMap).toMatchSnapshot();

    expect(failures[1].validatorName).toBe('NamespacesNamesMustNotHaveOnlyDifferentCasing');
    expect(failures[1].category).toBe('error');
    expect(failures[1].message).toMatchSnapshot();
    expect(failures[1].sourceMap).toMatchSnapshot();
  });
});
