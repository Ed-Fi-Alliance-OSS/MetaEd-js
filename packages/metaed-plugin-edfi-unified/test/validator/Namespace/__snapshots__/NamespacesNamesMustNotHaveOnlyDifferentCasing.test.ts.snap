// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when validating namespaces are different only in casing should have validation failure 1`] = `"EdFi is a project name that differs only in case from another name.  Make the casing the same to indicate the same namespace or choose a different name."`;

exports[`when validating namespaces are different only in casing should have validation failure 2`] = `null`;

exports[`when validating namespaces are different only in casing should have validation failure 3`] = `"EDFI is a project name that differs only in case from another name.  Make the casing the same to indicate the same namespace or choose a different name."`;

exports[`when validating namespaces are different only in casing should have validation failure 4`] = `null`;
